### PR TITLE
feat(Core/Database): port TrinityCore database API

### LINKED ISSUE #4669: Move/Port TrinityCore Database API [$165]
## Fix more memleaks
- https://github.com/TrinityCore/TrinityCore/commit/53482f729995889936ad8fba9fc5789d9cdba85
- https://github.com/TrinityCore/TrinityCore/commit/6b8f3d0f94b7a611eba81b7a2ff476cf75d09f4a

## Use new callback api 
- https://github.com/TrinityCore/TrinityCore/commit/4c27203c8f36dd2a5df0a4ae69fbdc4c9140b29d

## Completely remove ace inherited (in database)
- Part #3451

## Use api for next PR's
- #3352

## And more more more
- ...

<bountysource-plugin>

---
There is a **[$165 open bounty](https://www.bountysource.com/issues/96848456-move-port-trinitycore-database-api?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## Changes Proposed:
- It's about time, hah
- Port all changes TrinityCore database

## Issues Addressed:
- Closes #4669
- Part #3739 (code only, start support after add boost suport)
- Maybe #4005
- Maybe #3679

## Tests Performed:
- Char (Loading/create/delete)
- Hunter pets (loading, combat, stable master, rename)

## How to Test the Changes:
- Check char create/rename/customize/delete/loading and other
- Check all pets functions (rename/add) and other
- Check character save system

## Known Issues and TODO List:
- [x] License
- [ ] https://github.com/ElunaLuaEngine/Eluna/pull/358
- [ ] https://github.com/azerothcore/mod-transmog/pull/45
- [x] https://github.com/azerothcore/azerothcore-wotlk/pull/5671
- [x] #5950
- [x] Fix `clang-7`
- [ ] change wh
- [x] `.char rename`
- [x] `.char customize`

## Target Branch(es):
- [x] Master
 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by FrancescoBorzi** [comment]
> Ready to test without modules with db

not yet, the build is still failing (including one without modules)

**Comment by Winfidonarleyan** [comment]
> > Ready to test without modules with db
> 
> not yet, the build is still failing (including one without modules)

Clang 6 build failing

File: src/server/database/Database/DatabaseLoader.cpp
Comment (Kitzunu) [inline src/server/database/Database/DatabaseLoader.cpp]: why space?

Comment (Winfidonarleyan) [reply src/server/database/Database/DatabaseLoader.cpp]: Outdated

**Comment by Winfidonarleyan** [comment]
After merging this i start #3739

**Comment by Winfidonarleyan** [comment]
try build `clang7` in `Ubunutu 20.04`
https://gist.github.com/Winfidonarleyan/0ebeac8f3591ac825a915b1dce91f2d0

**Comment by Winfidonarleyan** [comment]
@azerothcore/testers @azerothcore/Developers 
We need your help.

Need to build this PR in `Debian` with `clang7`

**Comment by Sedatyf** [comment]
What's ``clang7``?

**Comment by Winfidonarleyan** [comment]
> What's `clang7`?

https://llvm.org/

**Comment by FrancescoBorzi** [comment]
I don't think we need to test this on debian with clang7, we might even deprecate clang7 if it causes too many issues (we recommend clang 10+ anyway).

We just need to have this tested asap, cc @azerothcore/testers

**Comment by Sedatyf** [comment]
I'll have a look at it after work

**Comment by necropola** [comment]
Neweest clang version in debian stable is still clang-7, buster-backports has clang-8. You need backports for cmake anyway since a few weeks ago.  You are basically about to depeicate debian stable entirely.

But that's just how developers roll these days.

**Comment by Sedatyf** [comment]
I got an issue when I want to launch servers (auth and world). Here the message I got:
```
DatabaseWorkerPool.cpp:51 in DatabaseWorkerPool<class WorldDatabaseConnection>::DatabaseWorkerPool FATAL ERROR:
Used MySQL library version (6.1.11 id 60111) does not match the version id used to compile WarheadCore (id 50716)
```
I have MySQL 5.7 and a MySQL connecter 6.1, do I have to change something on my side?

**Comment by Winfidonarleyan** [comment]
> I got an issue when I want to launch servers (auth and world). Here the message I got:
> 
> ```
> DatabaseWorkerPool.cpp:51 in DatabaseWorkerPool<class WorldDatabaseConnection>::DatabaseWorkerPool FATAL ERROR:
> Used MySQL library version (6.1.11 id 60111) does not match the version id used to compile WarheadCore (id 50716)
> ```
> 
> I have MySQL 5.7 and a MySQL connecter 6.1, do I have to change something on my side?

I'll take a look tomorrow

**Comment by Winfidonarleyan** [comment]
@Sedatyf I have considered your question. This no error, this сhecking for the same used libraries, so that there are no errors in the future
```cpp
WPFatal(mysql_get_client_version() == MYSQL_VERSION_ID, "Used MySQL library version (%s id %lu) does not match the version id used to compile AzerothCore (id %u)", 
mysql_get_client_info(), mysql_get_client_version(), MYSQL_VERSION_ID);
```

**Comment by Sedatyf** [comment]
Well in my case, I couldn't launch my servers. After this message, servers closed themselves 
I'll give a it try again tonight

**Comment by Sedatyf** [comment]
There is a conflict with latest push @Winfidonarleyan 
I can't pull your PR

**Comment by Winfidonarleyan** [comment]
> There is a conflict with latest push @Winfidonarleyan
> I can't pull your PR

1 hour for fix

**Comment by Winfidonarleyan** [comment]
@Sedatyf Done

**Comment by FrancescoBorzi** [comment]
what is the status of this? if I understand correctly, this is needed for other PRs as well as for completely remove ACE, so it should be high priority

**Comment by Winfidonarleyan** [comment]
> what is the status of this? if I understand correctly, this is needed for other PRs as well as for completely remove ACE, so it should be high priority

Yes, this need for remove ACE

**Comment by Winfidonarleyan** [comment]
Is someone testing this out?

**Comment by Sedatyf** [comment]
Yes I will test it asap

**Comment by Winfidonarleyan** [comment]
> Yes I will test it asap

okay, thanks

**Comment by Sedatyf** [comment]
Hmmm... It seems I still have a problem to launch ``authserver`` and ``worldserver``
I can build everything with no problem, but I have this when I'm trying to launch my server:
```
azerothcore\src\server\database\Database\DatabaseWorkerPool.cpp:51 in DatabaseWorkerPool<class WorldDatabaseConnection>::DatabaseWorkerPool 
FATAL ERROR: Used MySQL library version (6.1.11 id 60111) does not match the version id used to compile AzerothCore (id 50716)  
````

**Comment by Winfidonarleyan** [comment]
> Hmmm... It seems I still have a problem to launch `authserver` and `worldserver`
> I can build everything with no problem, but I have this when I'm trying to launch my server:
> 
> ```
> azerothcore\src\server\database\Database\DatabaseWorkerPool.cpp:51 in DatabaseWorkerPool<class WorldDatabaseConnection>::DatabaseWorkerPool 
> FATAL ERROR: Used MySQL library version (6.1.11 id 60111) does not match the version id used to compile AzerothCore (id 50716)  
> ```

Need use same lib for compile and startup apps

**Comment by Winfidonarleyan** [comment]
@Sedatyf Why are you using 2 different databases?

**Comment by Sedatyf** [comment]
I'm not actually. I'm using MySQL 5.7 and a MySQL connecter 6.1.
But I have only one database

**Comment by Winfidonarleyan** [comment]
> I'm not actually. I'm using MySQL 5.7 and a MySQL connecter 6.1.
> But I have only one database

Which one do you use?

**Comment by Sedatyf** [comment]
I don't understand what you're asking 😓

**Comment by Winfidonarleyan** [comment]
> I don't understand what you're asking 😓

You can use one db, example mariadb

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan he is using one DB, which is MySQL

**Comment by Winfidonarleyan** [comment]
> @Winfidonarleyan he is using one DB, which is MySQL

Need using this lib for compile and for startup

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan which lib?

**Comment by Winfidonarleyan** [comment]
> @Winfidonarleyan which lib?

libmysql.dll

File: src/server/database/Database/Field.cpp
Comment (claudiodfc) [inline src/server/database/Database/Field.cpp:26]: Warhead? Another core to make Azerothcore even more confusing?

**Comment by claudiodfc** [comment]
Not a single line of useful comments, really makes you wonder who is going to maintain this code.

File: src/server/database/Database/Field.cpp
Comment (Winfidonarleyan) [reply src/server/database/Database/Field.cpp:26]: my bad

**Comment by Winfidonarleyan** [comment]
Any news?

**Comment by Sedatyf** [comment]
Did you change something on your side? Because if no, I didn't change anything in mine, so it will be the same

**Comment by Winfidonarleyan** [comment]
Just use same lib, you can?

**Comment by Winfidonarleyan** [comment]
@Sedatyf can show Cmake log?

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan if it worked before with different libs, why should it be broken now?

**Comment by Kitzunu** [comment]
This is nothing new. You should never use a lib of another version from what you have compiled with. 
```
Note: You need the exact version of libmysql to correspond to the MySQL you have downloaded. Due to this you cannot download the DLL from the web and need to take it out of the folder.

https://www.azerothcore.org/wiki/windows-core-installation
```

It would be like if I used 5.7 lib when I run an 8.0 server 

And in this case, according to https://github.com/azerothcore/azerothcore-wotlk/pull/5611#issuecomment-858609210, he has compiled using MySQL 5.7 but trying to use a 6.1 lib

**Comment by Sedatyf** [comment]
Hmmm... Okay. I'll find a new connector which match with my SQL version. But I don't understand why it's working on every other PR and not this

**Comment by Winfidonarleyan** [comment]
> Hmmm... Okay. I'll find a new connector which match with my SQL version. But I don't understand why it's working on every other PR and not this

Just use mysql 5.7, no need use connector for core o.o

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan what is the status of this?

**Comment by Sedatyf** [comment]
> @Winfidonarleyan what is the status of this?

I didn't tested again yet. As I don't understand why I don't need connector. If so, why it was mentioned and used in the installation process?

**Comment by Sedatyf** [comment]
But I didn't search for a connector with same version as MySQL. To be honest, if someone else than me can take a look at this for testing, and see if he/she has same problem as mine could help

**Comment by Winfidonarleyan** [comment]
@Sedatyf just use MariaDB https://downloads.mariadb.org/mariadb/10.5.10/

**Comment by Winfidonarleyan** [comment]
Ready

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan the build is failing

```
src/server/game/Handlers/VoiceChatHandler.cpp:14:5: fatal error: use of undeclared identifier 'LOG_DEBUG'
    LOG_DEBUG("network", "WORLD: CMSG_VOICE_SESSION_ENABLE");
```

**Comment by Winfidonarleyan** [comment]
> @Winfidonarleyan the build is failing
> 
> ```
> src/server/game/Handlers/VoiceChatHandler.cpp:14:5: fatal error: use of undeclared identifier 'LOG_DEBUG'
>     LOG_DEBUG("network", "WORLD: CMSG_VOICE_SESSION_ENABLE");
> ```

Okay

**Comment by Winfidonarleyan** [comment]
So, ready again

**Comment by FrancescoBorzi** [comment]
> @Sedatyf just use MariaDB https://downloads.mariadb.org/mariadb/10.5.10/

@Winfidonarleyan I don't get why he should use MariaDB. His setup with MySQL works already on AC master and, if I understood correctly, only with your PR it does not work. So this is already a testing feedback: the PR should not break the existing behaviour with MySQL. Can you please make sure it works for all users, regardless of their MySQL or MariaDB version?

**Comment by Winfidonarleyan** [comment]
> > @Sedatyf just use MariaDB https://downloads.mariadb.org/mariadb/10.5.10/
> 
> @Winfidonarleyan I don't get why he should use MariaDB. His setup with MySQL works already on AC master and, if I understood correctly, only with your PR it does not work. So this is already a testing feedback: the PR should not break the existing behaviour with MySQL. Can you please make sure it works for all users, regardless of their MySQL or MariaDB version?

if use same lib for compile and run - pr work normal

File: src/server/database/Database/DatabaseWorkerPool.cpp
Comment (Kitzunu) [inline src/server/database/Database/DatabaseWorkerPool.cpp:51]: @FrancescoBorzi @Sedatyf 
Here is what is breaking it for you. And this is intentional as having a lib version != the version of MySQL that was used to compile the core could cause issues.

This is good behavior to prevent issues that could come. It only makes sense.

https://github.com/azerothcore/azerothcore-wotlk/pull/5611#issuecomment-860242128

Comment (Kitzunu) [reply src/server/database/Database/DatabaseWorkerPool.cpp:51]: I will say again, it would be if I used MySQL library from 5.7 when I have compiled with MySQL 8.0, it does not make sense and the installation guide also points this out.

**Comment by Kitzunu** [comment]
Tests performed:

- 🟢 Created a character -> Works
- 🔴 .char rename -> When you log out and go to rename, it does not let you go past the "Please enter a new name" window
- 🔴 .char customize -> You are allowed to customize and change name, but the flag is not removed so you can customize your character infinitely.

File: src/server/database/Database/DatabaseWorkerPool.cpp
Comment (Kitzunu) [reply src/server/database/Database/DatabaseWorkerPool.cpp:51]: I compiled myself, got the same error, followed the libmysql.dll part of https://www.azerothcore.org/wiki/windows-core-installation#compiling-the-source and now it works :)

Comment (FrancescoBorzi) [reply src/server/database/Database/DatabaseWorkerPool.cpp:51]: makes sense, thanks for the clarification @Kitzunu 

cc @Sedatyf

**Comment by FrancescoBorzi** [comment]
> Tests performed:
> 
> * 🟢 Created a character -> Works
> * 🔴 .char rename -> When you log out and go to rename, it does not let you go past the "Please enter a new name" window
> * 🔴 .char customize -> You are allowed to customize and change name, but the flag is not removed so you can customize your character infinitely.

thanks for testing @Kitzunu . Please fix @Winfidonarleyan

**Comment by Winfidonarleyan** [comment]
> Tests performed:
> 
> * 🟢 Created a character -> Works
> * 🔴 .char rename -> When you log out and go to rename, it does not let you go past the "Please enter a new name" window
> * 🔴 .char customize -> You are allowed to customize and change name, but the flag is not removed so you can customize your character infinitely.

Big thanks for testing, I'm getting down to fixing

**Comment by Kitzunu** [comment]
Tests performed:

- 🟢 Created a character -> Works
- 🟢 .char rename -> Works
- 🟢 .char customize -> Works
- 🟢 .char changerace -> Works (Even though you have mail, but probably not an issue of this PR)
- 🟢 .char changefaction -> Works
- 🟢 Talents -> Works
- 🟢 Glyphs -> Works
- 🟢 Restarting server -> Works, everything is saved as I left it
- 🟢 Deleting character -> Works, It's deleted
- 🟢 Tamed beast -> Works
- 🟢 Revive Pet -> Works
- 🟢 Rename Pet -> Works
- 🟢 Stable -> Works
- 🟢 Pet talents -> Works
- 🟢 Restarting server -> Works, Pet still stabled, has new name.
- 🟢 Out of Stable -> Works

To me, all looks good =)

**Review by Kitzunu — APPROVED** [review]
I won't be reviewing 156 files, putting my trust in the CI and my tests

**Comment by Winfidonarleyan** [comment]
> Tests performed:
> 
> * 🟢 Created a character -> Works
> * 🟢 .char rename -> Works
> * 🟢 .char customize -> Works
> * 🟢 .char changerace -> Works (Even though you have mail, but probably not an issue of this PR)
> * 🟢 .char changefaction -> Works
> * 🟢 Talents -> Works
> * 🟢 Glyphs -> Works
> * 🟢 Restarting server -> Works, everything is saved as I left it
> * 🟢 Deleting character -> Works, It's deleted
> * 🟢 Tamed beast -> Works
> * 🟢 Revive Pet -> Works
> * 🟢 Rename Pet -> Works
> * 🟢 Stable -> Works
> * 🟢 Pet talents -> Works
> * 🟢 Restarting server -> Works, Pet still stabled, has new name.
> * 🟢 Out of Stable -> Works
> 
> To me, all looks good =)

Very good, big thanks for the test

**Comment by Helias** [comment]
merge conflicts again :S
let's merge this big PR when we can @Kitzunu @FrancescoBorzi

**Comment by jackpoz** [comment]
This PR removed `* Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/>` from a few ported files. Was that supposed to happen ?

**Comment by Kitzunu** [comment]
cc @Winfidonarleyan please take a look at this

**Comment by Winfidonarleyan** [comment]
> This PR removed `* Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/>` from a few ported files. Was that supposed to happen ?

I fix today, sorry

**Comment by Winfidonarleyan** [comment]
@jackpoz, https://github.com/azerothcore/azerothcore-wotlk/commit/52aa0e996640dafe2b66d2e68d98ec68f9d9aa5e

**Pull request merged**
